pkg/hydrate/controller.go

```diff
@@ -115,13 +115,28 @@ func (h *Hydrator) runHydrate(sourceCommit, syncDir string) HydrationError {
 	if err := kustomizeBuild(syncDir, dest, true); err != nil {
 		return err
 	}
+	if changed, commit := h.checkCommitChanged(sourceCommit); changed {
+		return NewInternalError(fmt.Errorf("source commit changed while running Kustomize build, was %s, now %s. It will be retried in the next sync", commit, sourceCommit))
+	}
+
 	if err := updateSymlink(h.HydratedRoot.OSPath(), h.HydratedLink, newHydratedDir.OSPath()); err != nil {
 		return NewInternalError(errors.Wrapf(err, "unable to update the symbolic link to %s", newHydratedDir.OSPath()))
 	}
 	klog.Infof("Successfully rendered %s for commit %s", syncDir, sourceCommit)
 	return nil
 }
 
+// checkCommitChange returns true if the computed commit is different from given commit
+func (h *Hydrator) checkCommitChanged(commit string) (bool, string) {
+	absSourceDir := h.SourceRoot.Join(cmpath.RelativeSlash(h.SourceLink))
+	newCommit, _, err := SourceCommitAndDir(h.SourceType, absSourceDir, h.SyncDir, h.ReconcilerName)
+	if err != nil {
+		klog.Errorf("failed to get the commit hash and sync directory from the source directory %s: %v", absSourceDir.OSPath(), err)
+		return false, ""
```

Comment:
In the error case, you return false for commit unchanged, which might result in proceeding with a new commit.
Instead of logging the error, I would suggest returning the error.

Reply:
Added error in return and check